File: easy/good_pairs/good_pairs.py
class Solution(object):
    def numIdenticalPairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        total_pairs = 0
        values_processed = {}

        for i in nums:
            if not i in values_processed:
                occurrences = nums.count(i)
                if occurrences > 1:
                    all_pairs = occurrences * (occurrences - 1)
                    increasing_order_pairs = all_pairs / 2
                    total_pairs = total_pairs + increasing_order_pairs
                    values_processed[i] = True

        return int(total_pairs)


s = Solution()
print("Solution 1 : ", s.numIdenticalPairs([1, 2, 3, 1, 1, 3]))
print("Solution 2 : ", s.numIdenticalPairs([1, 1, 1, 1]))
print("Solution 3 : ", s.numIdenticalPairs([1, 2, 3]))

File: easy/good_pairs/good_pairs.go
package main

import "fmt"

func countPairs(valToNumberOfOccurrences map[int]int) int {
	totalPairs := 0
	for _, occurrences := range valToNumberOfOccurrences {
		allPairs := occurrences * (occurrences - 1)
		increasingOrderPairs := allPairs / 2
		totalPairs = totalPairs + increasingOrderPairs
	}

	return totalPairs
}

func countOccerrences(nums []int) map[int]int {
	valToNumberOfOccurrences := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		if _, ok := valToNumberOfOccurrences[nums[i]]; ok {
			valToNumberOfOccurrences[nums[i]]++
		} else {
			valToNumberOfOccurrences[nums[i]] = 1
		}
	}

	return valToNumberOfOccurrences
}

func numIdenticalPairs(nums []int) int {
	valToNumberOfOccurrences := countOccerrences(nums)
	return countPairs(valToNumberOfOccurrences)
}

func main() {
	fmt.Println("Solution 1 : ", numIdenticalPairs([]int{1, 2, 3, 1, 1, 3}))
	fmt.Println("Solution 2 : ", numIdenticalPairs([]int{1, 1, 1, 1}))
	fmt.Println("Solution 3 : ", numIdenticalPairs([]int{1, 2, 3}))
}
